### PR TITLE
smtplib: Improve bytes handling

### PR DESCRIPTION
**Pull request by JelleZijlstra**

### CODE REVIEW

File: stdlib/smtplib.pyi [hunk 1]
```diff
@@ -29,7 +29,7 @@ __all__ = [
 _Reply: TypeAlias = tuple[int, bytes]
 _SendErrs: TypeAlias = dict[str, _Reply]
 # Should match source_address for socket.create_connection
```
Comment (AlexWaygood) [inline stdlib/smtplib.pyi:31]: This comment looks out of date — looks like the alias has diverged somewhat from https://github.com/python/typeshed/blob/9f0d0b9af7d4980431bbd62ca38497cf33308093/stdlib/_socket.pyi#L20

Comment (JelleZijlstra) [reply stdlib/smtplib.pyi:31]: We should just reuse that alias, not sure why I missed that. Source address just gets passed to `socket.create_connection`.

File: stdlib/smtplib.pyi [hunk 2]
```diff
@@ -9,6 +9,8 @@ from types import TracebackType
 from typing import Any, Protocol, overload
 from typing_extensions import TypeAlias
 
+from _socket import _Address
```
Comment (AlexWaygood) [inline stdlib/smtplib.pyi:12]: To keep it clear what the `_Address` alias is referring to, maybe:

```suggestion
from _socket import _Address as _SocketAddress
```

Or

```suggestion
from _socket import _Address as _SourceAddress
```

File: stdlib/smtplib.pyi [hunk 3]
```diff
@@ -9,6 +9,8 @@ from types import TracebackType
 from typing import Any, Protocol, overload
 from typing_extensions import TypeAlias
 
+from _socket import _Address as _SourceAddress
```
Comment (AlexWaygood) [inline stdlib/smtplib.pyi:12]: Oh, we should add `_socket` to this list, so isort recognises it as part of the stdlib:

https://github.com/python/typeshed/blob/9f0d0b9af7d4980431bbd62ca38497cf33308093/pyproject.toml#L15

Comment (JelleZijlstra) [reply stdlib/smtplib.pyi:12]: I can do that in a separate PR.

File: stdlib/smtplib.pyi [hunk 4]
```diff
@@ -125,16 +125,16 @@ class SMTP:
     @overload
     def auth_cram_md5(self, challenge: None = ...) -> None: ...
     @overload
-    def auth_cram_md5(self, challenge: bytes) -> str: ...
-    def auth_plain(self, challenge: bytes | None = ...) -> str: ...
-    def auth_login(self, challenge: bytes | None = ...) -> str: ...
+    def auth_cram_md5(self, challenge: ReadableBuffer) -> str: ...
+    def auth_plain(self, challenge: ReadableBuffer | None = ...) -> str: ...
```
Comment (AlexWaygood) [inline stdlib/smtplib.pyi:129]: The `challenge` parameter doesn't look like it's actually used for this one

**Review by AlexWaygood — APPROVED** [review]

File: stdlib/smtplib.pyi (hunk 4 above)
Comment (JelleZijlstra) [reply stdlib/smtplib.pyi:129]: The next one doesn't really use it either, I figured I should be consistent with `auth_cram_md5`.

Comment (AlexWaygood) [reply stdlib/smtplib.pyi:129]: Yeah, that makes sense, I figured as much!

File: stdlib/smtplib.pyi (hunk 3 above)
Comment (JelleZijlstra) [reply stdlib/smtplib.pyi:12]: #9098

**Pull request merged**